File: peacecheejecake/leetcode/121_Best_Time_To_Buy_And_Sell_Stock.py
# https://leetcode.com/problems/best-time-to-buy-and-sell-stock/

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        max_prices = [0] * len(prices)
        max_profit = 0
        for i in range(len(prices) - 2, -1, -1):
            max_after = max(max_prices[i + 1], prices[i + 1])
            max_prices[i] = max_after
            max_profit = max(max_profit, max_after - prices[i])
        return max_profit
        

File: sally/leetcode/121_best-time-to-buy-and-sell-stock.cpp
// 121. Best Time to Buy and Sell Stock (Easy)
// Runtime: 96 ms, faster than 98.13% of C++ online submissions for Best Time to Buy and Sell Stock.
// Memory Usage: 93.4 MB, less than 53.95% of C++ online submissions for Best Time to Buy and Sell Stock.

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int N = prices.size();
        int answer = 0;
        int min_val = prices[0];
        for(int i = 0; i < N; i++){
            answer = max(answer, prices[i]-min_val);
            min_val = min(min_val, prices[i]);
        }
        return answer;
    }
};

// N^2 time limit